From 8071ffe49cfa489d356ab4c924d6c7f9b86765c1 Mon Sep 17 00:00:00 2001
From: Adithya R <gh0strider.2k18.reborn@gmail.com>
Date: Thu, 8 Sep 2022 16:30:48 +0530
Subject: [PATCH 6/6] PIHooks: Spoof device as Pixel XL for Google Photos

Enables unlimited photos and videos backup (restored
functionality from PixelPropsUtils).

Make this optional via overlay, as it needs to be
disabled on some devices such as Pixel 6 series
in order to utilize its tensor features.

While we're at it, move the feature blacklist here.

Change-Id: I1b173a0fd2e0f5cef946339c1a121e59fb5b430d
Signed-off-by: Nishant Kumar <www.rajsonu13@gmail.com>
Signed-off-by: penglezos <panagiotisegl@gmail.com>
---
 .../app/ApplicationPackageManager.java        |  4 +-
 .../internal/util/PropImitationHooks.java     | 42 +++++++++++++++++++
 core/res/res/values/custom_config.xml         |  2 +
 core/res/res/values/custom_symbols.xml        |  3 ++
 4 files changed, 50 insertions(+), 1 deletion(-)

diff --git a/core/java/android/app/ApplicationPackageManager.java b/core/java/android/app/ApplicationPackageManager.java
index 475e0f4cb05b..313a67d00c2a 100644
--- a/core/java/android/app/ApplicationPackageManager.java
+++ b/core/java/android/app/ApplicationPackageManager.java
@@ -120,6 +120,7 @@ import com.android.internal.annotations.GuardedBy;
 import com.android.internal.annotations.Immutable;
 import com.android.internal.annotations.VisibleForTesting;
 import com.android.internal.os.SomeArgs;
+import com.android.internal.util.PropImitationHooks;
 import com.android.internal.util.UserIcons;
 
 import com.nvidia.NvAppProfileService;
@@ -820,7 +821,8 @@ public class ApplicationPackageManager extends PackageManager {
 
     @Override
     public boolean hasSystemFeature(String name, int version) {
-        return mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version));
+        return PropImitationHooks.hasSystemFeature(name,
+                mHasSystemFeatureCache.query(new HasSystemFeatureQuery(name, version)));
     }
 
     /** @hide */
diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index 54ad67da524d..65c8d1bf3059 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -25,6 +25,8 @@ import com.android.internal.R;
 
 import java.lang.reflect.Field;
 import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
 
 public class PropImitationHooks {
 
@@ -42,8 +44,35 @@ public class PropImitationHooks {
     private static final String PACKAGE_GMS = "com.google.android.gms";
     private static final String PROCESS_GMS_UNSTABLE = PACKAGE_GMS + ".unstable";
 
+    private static final String PACKAGE_GPHOTOS = "com.google.android.apps.photos";
+    private static final Map<String, Object> sP1Props = new HashMap<>();
+    static {
+        sP1Props.put("BRAND", "google");
+        sP1Props.put("MANUFACTURER", "Google");
+        sP1Props.put("DEVICE", "marlin");
+        sP1Props.put("PRODUCT", "marlin");
+        sP1Props.put("MODEL", "Pixel XL");
+        sP1Props.put("FINGERPRINT", "google/marlin/marlin:10/QP1A.191005.007.A3/5972272:user/release-keys");
+    }
+    private static final String[] sFeaturesBlacklist = {
+        "PIXEL_2017_EXPERIENCE",
+        "PIXEL_2017_PRELOAD",
+        "PIXEL_2018_PRELOAD",
+        "PIXEL_2019_EXPERIENCE",
+        "PIXEL_2019_MIDYEAR_EXPERIENCE",
+        "PIXEL_2019_MIDYEAR_PRELOAD",
+        "PIXEL_2019_PRELOAD",
+        "PIXEL_2020_EXPERIENCE",
+        "PIXEL_2020_MIDYEAR_EXPERIENCE",
+        "PIXEL_2021_EXPERIENCE",
+        "PIXEL_2021_MIDYEAR_EXPERIENCE"
+    };
+    private static final boolean sSpoofPhotos =
+            Resources.getSystem().getBoolean(R.bool.config_spoofGooglePhotos);
+
     private static volatile boolean sIsGms = false;
     private static volatile boolean sIsFinsky = false;
+    private static volatile boolean sIsPhotos = false;
 
     public static void setProps(Application app) {
         final String packageName = app.getPackageName();
@@ -55,6 +84,7 @@ public class PropImitationHooks {
 
         sIsGms = packageName.equals(PACKAGE_GMS) && processName.equals(PROCESS_GMS_UNSTABLE);
         sIsFinsky = packageName.equals(PACKAGE_FINSKY);
+        sIsPhotos = sSpoofPhotos && packageName.equals(PACKAGE_GPHOTOS);
 
         /* Set Certified Fingerprint for GMSCore or Finsky
          * Set Stock Fingerprint for ARCore
@@ -66,6 +96,9 @@ public class PropImitationHooks {
         } else if (!sStockFp.isEmpty() && packageName.equals(PACKAGE_ARCORE)) {
             dlog("Setting stock fingerprint for: " + packageName);
             setPropValue("FINGERPRINT", sStockFp);
+        } else if (sIsPhotos) {
+            dlog("Spoofing Pixel XL for Google Photos");
+            sP1Props.forEach((k, v) -> setPropValue(k, v));
         }
     }
 
@@ -94,6 +127,15 @@ public class PropImitationHooks {
         }
     }
 
+    public static boolean hasSystemFeature(String name, boolean def) {
+        if (sIsPhotos && def &&
+                Arrays.stream(sFeaturesBlacklist).anyMatch(name::contains)) {
+            dlog("Blocked system feature " + name + " for Google Photos");
+            return false;
+        }
+        return def;
+    }
+
     public static void dlog(String msg) {
       if (DEBUG) Log.d(TAG, msg);
     }
diff --git a/core/res/res/values/custom_config.xml b/core/res/res/values/custom_config.xml
index 70b7fa082814..603cc9e17de9 100644
--- a/core/res/res/values/custom_config.xml
+++ b/core/res/res/values/custom_config.xml
@@ -6,4 +6,6 @@
     <!-- Fingerprint from stock ROM -->
     <string name="config_stockFingerprint" translatable="false"></string>
 
+    <!-- Whether to spoof device as Pixel XL for Google Photos -->
+    <bool name="config_spoofGooglePhotos">true</bool>
 </resources>
diff --git a/core/res/res/values/custom_symbols.xml b/core/res/res/values/custom_symbols.xml
index 3bb4834c99f4..23694d172326 100644
--- a/core/res/res/values/custom_symbols.xml
+++ b/core/res/res/values/custom_symbols.xml
@@ -5,4 +5,7 @@
 
     <!-- Fingerprint from stock ROM -->
     <java-symbol type="string" name="config_stockFingerprint" />
+
+    <!-- Whether to spoof device as Pixel XL for Google Photos -->
+    <java-symbol type="bool" name="config_spoofGooglePhotos" />
 </resources>
-- 
2.39.1

